librz/bin/bin.c

```diff
@@ -1316,188 +1317,74 @@ RZ_API const RzBinXtrPlugin *rz_bin_xtrplugin_get(RZ_NONNULL RzBin *bin, RZ_NONN
 }
 
-static char *bin_demangle_cxx(RzBinFile *bf, const char *symbol, ut64 vaddr) {
-	char *out = rz_demangler_cxx(symbol);
-	if (!out || !bf) {
-		return out;
-	}
-	char *sign = (char *)strchr(out, '(');
-	if (!sign) {
-		return out;
-	}
-
-	char *str = out;
-	char *ptr = NULL;
-	char *method_name = NULL;
-	for (;;) {
-		ptr = strstr(str, "::");
-		if (!ptr || ptr > sign) {
-			break;
-		}
-		method_name = ptr;
-		str = ptr + 1;
-	}
-
-	if (RZ_STR_ISEMPTY(method_name)) {
-		return out;
-	}
-
-	*method_name = 0;
-	RzBinSymbol *sym = rz_bin_file_add_method(bf, out, method_name + 2, 0);
-	if (sym) {
-		if (sym->vaddr != 0 && sym->vaddr != vaddr) {
-			RZ_LOG_INFO("Duplicated method found: %s\n", sym->name);
-		}
-		if (sym->vaddr == 0) {
-			sym->vaddr = vaddr;
-		}
-	}
-	*method_name = ':';
-	return out;
-}
-
-static char *bin_demangle_rust(RzBinFile *bf, const char *symbol, ut64 vaddr) {
-	char *demangled = rz_demangler_rust(symbol);
-	if (!demangled) {
-		return demangled;
-	}
-
-	char *str = demangled;
-	char *ptr = NULL;
-	char *name = NULL;
-	while ((ptr = strstr(str, "::"))) {
-		name = ptr;
-		str = ptr + 2;
-	}
-
-	if (!name || RZ_STR_ISEMPTY(name + 2)) {
-		return demangled;
-	}
-
-	*name = 0;
-	RzBinSymbol *sym = rz_bin_file_add_method(bf, demangled, name + 2, 0);
-	if (sym && sym->vaddr == 0) {
-		sym->vaddr = vaddr;
-	}
-	*name = ':';
-	return demangled;
-}
-
 /**
- * \brief Demangles a symbol based on the language or the RzBinFile data
+ * \brief Demangles a symbol based on the language or by iterating all demanglers.
  *
- * This function demangles a symbol based on the language or the RzBinFile data
- * When C++ or rust is selected as the language, it will add methods into the
- * RzBinFile structure based on the demangled symbol.
- * When libs is set to true, the demangled symbol will be appended to the
- * library name <libname>_<demangled symbol>.
+ * This function demangles a symbol based on the language or by iterating all demanglers.
+ * The iteration of all demanglers is available only if the RzBin pointer is not NULL.
  *
- * \param bf RzBinFile data to be used for demangling
- * \param language Language to be used for demanglind
- * \param symbol Symbol to be demangled
- * \param vaddr vaddr of the \p symbol to be demangled
- * \param libs Append the library name to the demangled symbol, if set to true
- * \return char* Demangled name of the \p symbol
+ * \param   bin       The RzBin context to used for demangling
+ * \param   language  The language to be used for demangling
+ * \param   mangled   The mangled string to be demangled
+ * \return  On success returns a valid pointer, otherwise NULL.
  */
-RZ_API RZ_OWN char *rz_bin_demangle(RZ_NULLABLE RzBinFile *bf, RZ_NULLABLE const char *language, RZ_NULLABLE const char *symbol, ut64 vaddr, bool libs) {
-	if (RZ_STR_ISEMPTY(symbol)) {
+RZ_API RZ_OWN char *rz_bin_demangle(RZ_NULLABLE RzBin *bin, RZ_NULLABLE const char *language, RZ_NULLABLE const char *mangled) {
```

Comment:
Do you think this can be aggregated with the functions in `bin_demangle.c`? I see many similarities.

Reply:
i can actually move it there. makes more sense.